azure-mgmt-resources/src/main/java/com/microsoft/azure/management/resources/fluentcore/utils/ProviderRegistrationInterceptor.java

```diff
@@ -0,0 +1,89 @@
+/**
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License. See License.txt in the project root for
+ * license information.
+ */
+
+package com.microsoft.azure.management.resources.fluentcore.utils;
+
+import com.microsoft.azure.AzureResponseBuilder;
+import com.microsoft.azure.CloudError;
+import com.microsoft.azure.credentials.AzureTokenCredentials;
+import com.microsoft.azure.management.resources.Provider;
+import com.microsoft.azure.management.resources.implementation.ResourceManager;
+import com.microsoft.azure.serializer.AzureJacksonAdapter;
+import com.microsoft.rest.RestClient;
+import okhttp3.Interceptor;
+import okhttp3.Response;
+import okhttp3.ResponseBody;
+import okio.Buffer;
+import okio.BufferedSource;
+
+import java.io.IOException;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * An interceptor for automatic provider registration in Azure.
+ */
+public class ProviderRegistrationInterceptor implements Interceptor {
+    private AzureTokenCredentials credentials;
+
+    /**
+     * Initialize a provider registration interceptor with a credential that's authorized
+     * to register the provider.
+     * @param credentials the credential for provider registration
+     */
+    public ProviderRegistrationInterceptor(AzureTokenCredentials credentials) {
+        this.credentials = credentials;
+    }
+
+    @Override
+    public Response intercept(Chain chain) throws IOException {
+        Response response = chain.proceed(chain.request());
+        if (!response.isSuccessful()) {
+            String content = errorBody(response.body());
+            RestClient restClient = new RestClient.Builder()
+                    .withBaseUrl("https://" + chain.request().url().host())
+                    .withCredentials(credentials)
+                    .withSerializerAdapter(new AzureJacksonAdapter())
+                    .withResponseBuilderFactory(new AzureResponseBuilder.Factory())
+                    .build();
```

Comment:
Given that `RestClient.Builder()...build()` is a heavy action, should we create RestClient only if `code` is `MissingSubscriptionRegistration`? so here we left with creation of `AzureJacksonAdapter`, that single instance can be used for both deserializing CloudError and with `withSerializerAdapter` wither?